### DISCUSSION TITLE
How to avoid `Exported variable ... has or is using name 'ZodFileDef' from external module ... but cannot be named.`?

### DISCUSSION THREAD

**Discussion by McMerph** [RobinTail/express-zod-api#951] [Q&A] [answered]
I'm building a library with custom endpoint factories to use in multiple projects.
But if I export an endpoint factory that uses `ez.file()`:
```
import { createResultHandler, EndpointsFactory, ez } from "express-zod-api";
import { Readable } from "stream";
import { z } from "zod";

const isNonNullObj = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null;

type EndpointOutput = { fileName: string; stream: Readable };
const isEndpointOutput = (v: unknown): v is EndpointOutput =>
  isNonNullObj(v) &&
  "fileName" in v &&
  typeof v["fileName"] === "string" &&
  "stream" in v &&
  v["stream"] instanceof Readable;

export const getFileStreamingEndpointsFactory = (mimeType: string) =>
  new EndpointsFactory(
    createResultHandler({
      getPositiveResponse: () => ({
        schema: ez.file().binary(),
        mimeType,
      }),
      getNegativeResponse: () => ({
        schema: z.string(),
        mimeType: "text/plain",
      }),
      handler({ response, error, output }) {
        if (error) {
          response.status(500).send(error.message);
          return;
        }

        if (!isEndpointOutput(output)) {
          response.status(500).send("Invalid endpoint output!");
          return;
        }

        response.header(
          "Content-Disposition",
          `attachment; filename="${output.fileName}"`
        );
        output.stream.pipe(response);
      },
    })
  );
```
then there is a `tsc` error:
```
TS4023: Exported variable 'getFileStreamingEndpointsFactory' has or is using name 'ZodFileDef' from external module "{{some path}}/node_modules/.pnpm/express-zod-api@10.0.0_n5zqmeauuo4jduap7lnbjtm454/node_modules/express-zod-api/dist/index" but cannot be named.
```

`tsconfig.json`:
```
{
  "extends": "@tsconfig/node18-strictest/tsconfig.json",
  "compilerOptions": {
    "isolatedModules": true,
    "esModuleInterop": true,
    "rootDir": "src",
    "outDir": "dist",
    "declaration": true,
    "declarationDir": "./distTypes",
    "sourceMap": true,
    "declarationMap": true
  },
  "include": ["**/*.ts", "**/*.js", ".eslintrc.js"],
  "exclude": [
    "coverage/",
    "dist/",
    "distTypes/",
    ".eslintrc.js",
    "jest.config.js"
  ]
}
```

How to avoid this error and what is its cause?

**Comment by RobinTail** [comment, marked as answer]
You can get rid of this error by disabling `declaration` in your tsconfig, @McMerph 

It's caused by the fact that the type ZodFileDef is not exported by `express-zod-api`, but your code is configured for generating `.d.ts` files, which in some way somehow need it in order to describe your entities.

**Reply by RobinTail** [reply]
Let me know, @McMerph , if generating declarations on files using `ez` namespace is unavoidable.

**Reply by McMerph** [reply]
I enabled `declaration` intentionally to provide intellisense to js code. So, I don't want to disable it.
However, I can use `z.string()` instead of `ez.file().binary()`.

> the type ZodFileDef is not exported by express-zod-api

Is it expected?

**Reply by RobinTail** [reply]
Hello @McMerph ,

> I enabled declaration intentionally to provide intellisense to js code.

I supposed that `Client` generator should be used for that purpose.

> So, I don't want to disable it.

Ok. I'll make it available then.

> Is it expected?

I [considered it as a redundancy here](https://github.com/RobinTail/express-zod-api/pull/864#issuecomment-1461057467) since having OpenAPI and Typescript generators.

**Reply by RobinTail** [reply]
I made an issue out of it, #952 .
Is it the only issue of that kind, @McMerph ?

**Reply by RobinTail** [reply]
@McMerph , the issue should be fix in v10.3.1 🚀 
Please try it and let me know if it's OK now.

**Reply by McMerph** [reply]
Still the same error in v10.3.1.

**Reply by RobinTail** [reply]
Is it about `ZodFileDef` or another entity, @McMerph ?

**Reply by McMerph** [reply]
Yes, it is about the `ZodFileDef`, exactly the same output

**Reply by RobinTail** [reply]
Hmmm, ok.
I'll try another approach then. Thank you for checking.

**Reply by RobinTail** [reply]
🚀 Please try v10.3.2, @McMerph

**Reply by McMerph** [reply]
Fixed! Thank you!
